bpo-41810: Reintroduce `types.EllipsisType`, `.NoneType` & `.NotImplementedType`

BvB93 opened the pull request:
See the relevant BPO for more details: https://bugs.python.org/issue41810.

<!-- issue-number: [bpo-41810](https://bugs.python.org/issue41810) -->
https://bugs.python.org/issue41810
<!-- /issue-number -->

Commit:
Reintroduced `types.EllipsisType`

Commit:
Replaced `type(...)` and `type(ellipsis)` with `types.EllipsisType`

Commit:
Added `EllipsisType` to the documentation

Commit:
Added Misc/NEWS.d and whatsnew entries

Commit:
Updated the acknowledgements

Commit:
Reintroduced `types.NoneType` and `types.NotImplementedType`

Commit:
Replaced `type(NotImplemented)` and `type(None)`

`type(NotImplemented)` and `type(None)` have, respectively, been replaced with `types.NotImplementedType` and `types.NoneType`

Commit:
Added `types.NoneType` & `types.NotImplementedType` to whatsnew and the docs

ericvsmith commented on Misc/NEWS.d/next/Library/2020-09-20-15-14-05.bpo-41810.7l8lyV.rst:
```diff
@@ -0,0 +1,2 @@
+:data:`types.EllipsisType`, :data:`types.EllipsisType` and
+:data:`types.NoneType` have been reintroduced.
```
We should mention the reason behind this reintroduction: for type checkers which can't use `type(Ellipsis)`, etc.

BvB93 replied:
I added a few words about the motivation in https://github.com/python/cpython/pull/22336/commits/370149882d2cea95f2fed7c732bcbb6ae343a3fe.

ericvsmith commented on Misc/NEWS.d/next/Library/2020-09-20-15-14-05.bpo-41810.7l8lyV.rst:
```diff
@@ -0,0 +1,2 @@
+:data:`types.EllipsisType`, :data:`types.EllipsisType` and
```
One of these should be `type.NotImplementedType`.

BvB93 replied:
Good catch; fixed in https://github.com/python/cpython/pull/22336/commits/f682623aa3001a4ac19e727c3103cdb0dce8672e.

ericvsmith reviewed (changes requested):
I've not done a thorough review yet. These comments are just about the blurb.

Commit:
Replaced a duplicate `EllipsisType` with `NotImplementedType`

Addresses https://github.com/python/cpython/pull/22336#discussion_r492317442

Commit:
Provide motivation for the type reintroduction

Addresses https://github.com/python/cpython/pull/22336#discussion_r492317123

ericvsmith commented on Misc/NEWS.d/next/Library/2020-09-20-15-14-05.bpo-41810.7l8lyV.rst:
```diff
@@ -1,2 +1,3 @@
-:data:`types.NoneType` have been reintroduced.
+:data:`types.EllipsisType`, :data:`types.NotImplementedType` and
+:data:`types.NoneType` have been reintroduced, aforementioned types
+now being useable as annotations.
```
I'm not trying to be overly picky here, but I really think you need to mention type checkers. If I was reading this comment 5 years from now, I'd think "but type(Ellipsis) was already usable as an annotation, so we should delete EllipsisType".

BvB93 replied:
Fair enough, fixed in https://github.com/python/cpython/pull/22336/commits/33caa6fb32ec98766a9402670176ddde16d47738.

Commit:
Explicitly mention static type checkers

Commit:
Revert the `type(...)` to `types....` replacements

Reverts https://github.com/python/cpython/pull/22336/commits/2791a239bacfe1f16e9672ca063cdb1f29e9a147 and https://github.com/python/cpython/pull/22336/commits/0a97a885178adaf79d93041aed5bbab3c9e7d264.
Save this for a follow up.

gvanrossum reviewed (commented):
I think the docs for the three special objects should not start with “The sole value of the type ...”.

The first sentence should serve as a brief summary of what the object is used for. You can then follow up with the detail about its singleton type.

Otherwise LGTM. Thanks for your patience!

ericvsmith reviewed (approved):
This looks good to me!

terryjreedy commented on Doc/library/constants.rst:
```diff
@@ -19,14 +19,15 @@ A small number of constants live in the built-in namespace.  They are:
 .. data:: None
 
-   The sole value of the type ``NoneType``.  ``None`` is frequently used to
+   The sole value of the type :data:`types.NoneType`. ``None`` is frequently used to
```
"The sole value of the type NoneType" is most easily read (parsed) as referring to NoneType's usefulness.  I expected it to be followed by 'is to ...' and initially read 'is frequently' as a grammatical error.  It took awhile to realize that this is a convoluted  way to say 'None'.  I strongly recommend a rewording.

```suggestion
   ''None'' is the only instance of :data:`types.NoneType`.  It is frequently used to
```
`types.NoneType` is already twice identified as a type, so a third mention is not needed.

BvB93 replied:
Based on your and Guido's feedback (https://github.com/python/cpython/pull/22336#pullrequestreview-493014262) I'm proposing something like this:

```
.. data:: None

   An object frequently used to represent the absence of a value, as when
   default arguments are not passed to a function. Assignments to ``None``
   are illegal and raise a :exc:`SyntaxError`.
   Sole instance of the :data:`types.NoneType` type.
```
Any thoughts?

ericvsmith replied:
That's much better, I think. Although I'd make the last sentence (with the correct markup): "None is the sole instance of the NoneType type."

terryjreedy commented on Doc/library/constants.rst:
```diff
@@ -19,14 +19,15 @@ A small number of constants live in the built-in namespace.  They are:
 
-   Special value which should be returned by the binary special methods
+   Sole value of the type :data:`types.NotImplementedType` and a
+   special value which should be returned by the binary special methods
```
See above.  Use a parallel construction.

```suggestion
   ``NotImplemented is the only instance of :data:`types.NotImplementedType`.  It is a
   special value which should be returned by the binary special methods
```

terryjreedy commented on Doc/library/constants.rst:
```diff
@@ -59,7 +60,8 @@ A small number of constants live in the built-in namespace.  They are:
 .. data:: Ellipsis
 
-   The same as the ellipsis literal "``...``".  Special value used mostly in conjunction
+   The same as the ellipsis literal "``...``" and sole value of the type
```
```suggestion
   The same as the ellipsis literal "``...``" and the only instance of
```

terryjreedy commented on Doc/whatsnew/3.10.rst:
```diff
@@ -145,6 +145,14 @@ Add :data:`sys.orig_argv` attribute: the list of the original command line
+
+Reintroduced the :data:`types.EllipsisType`, :data:`types.NoneType`
+and :data:`types.NotImplementedType` classes, providing a new set
+of types readily interpretable by type checkers.
```
I think this and the blurb are fine, providing a reason without saying either that this is the only reason or that these should be use everywhere.

terryjreedy reviewed (changes requested):
Approved, but only with misleading wording fixed.

Commit:
Reworded parts of the `constants` documentation

Addresses https://github.com/python/cpython/pull/22336#discussion_r492478512, https://github.com/python/cpython/pull/22336#discussion_r492479743 & https://github.com/python/cpython/pull/22336#discussion_r492480008
